week4/db-architecture-4.md

```diff
@@ -0,0 +1,88 @@
+
+Optimizer's functionality can be extended to:
+1. Select query plan space e.g. left deep plans only
+2. Estimate selectivity
+3. Decide on algorithms for searching the plan space
+4. Generate parallelizable plans
+5. Auto tuning based on query workload (e.g. choosing which views to materialize)
+
+#### Prepare and execute model
+   - Databases can store query plans for repeated execution, especially useful for queries with fixed structure over predictable data.
+   - Plans for dynamically built queries can be stored too and used when similar queries are submitted.
+   - There is an interesting tradeoff between predictable performance across invocations and optimal performance per invocation.
+
+### Query Executor
+   - Query executor on interpretable plans works on the iterator model (volcano-style). All relational operators are implemented as subclasses of the iterator class.
+
+#### Iterators
+   - Iterators couple dataflow with control flow i.e. a tuple is returned to a parent in the graph at the same time when control is returned. Which means a single DBMS thread is sufficient for executing the entire query graph.
+   - Single threaded iterators can be extended to support parallel query execution.
+
+#### Where is the data stored?
+   - In flight tuples are stored in Buffer Pool pages (BP tuples) or copied into memory heap (M tuples)
+   - Each iterator is allocated a fixed number of tuple descriptors (one for each input and one for output). A tuple descriptor is an array of column references where each reference is an actual memory address reference of a tuple and a column offset in that tuple.
+   - Using M tuples exclusively is a performance problem because copying data in and out of memory heap is a bottleneck, whereas directly referencing to BP pages results in pinning a page for a long time in BP memory which might cause page replacement issues for high loads.
+
+#### Handling DML queries
+
+   - SQL semantics dictate that a single SQL statement should not be able to see its own updates, careful implementation is needed to ensure this.
+   - ***The Halloween problem*** is an example of what can go wrong. The problem arises on statements like 
+      >give everyone whose salary is under $20K a 10% raise.” 
+   
+   - If a plan uses an index scan iterator over salary field piplined into an update iterator, it results in all low-paid employees receiving repeated raises until they earn more than $20K.
+   - One way to avoid this problem is not using indexes on the columns being updated, but this can be quite inefficient.
+   - Another solution is to store IDs of all tuples qualifying for updates into a temporary file and scanning the file to fetch tuples to be fed to the update operator. This works well when only a few tuples are to be updated because, in that case, the temporary file is small enough to fit entirely in the buffer pool.
```

Comment:
Solution to Halloween problem is not clear. It be added to QA file and docs for answering later.